File: fixtures/rust-sample/src/conditions.rs
// cgx-fixture: conditional and loop edges
// Covers: conditional edge (if/match arm), loop edge (for/while/loop body)

fn log_info(msg: &str) {
    println!("[INFO] {}", msg);
}

fn log_warn(msg: &str) {
    println!("[WARN] {}", msg);
}

fn log_error(msg: &str) {
    eprintln!("[ERROR] {}", msg);
}

fn process_item(x: i32) -> i32 {
    x * 2
}

fn cleanup() {
    log_info("cleanup called");
}

fn validate(x: i32) -> bool {
    x > 0
}

fn expensive_check(x: i32) -> bool {
    x % 2 == 0
}

/// Simple if/else — conditional edges to log_info and log_warn.
pub fn maybe_log(flag: bool) {
    if flag {
        log_info("flag is true");      // conditional edge: if-branch
    } else {
        log_warn("flag is false");     // conditional edge: else-branch
    }
}

/// match expression — each arm produces a conditional edge.
pub fn dispatch(code: u8) -> &'static str {
    match code {
        0 => { log_info("zero"); "zero" },      // conditional
        1 => { log_warn("one"); "one" },         // conditional
        _ => { log_error("other"); "other" },    // conditional
    }
}

/// for loop — loop edge on process_item calls inside the body.
pub fn process_all(items: &[i32]) -> Vec<i32> {
    let mut out = Vec::new();
    for &x in items {
        out.push(process_item(x));   // loop edge
    }
    out
}

/// while loop — loop edge.
pub fn count_down(mut n: u32) -> Vec<u32> {
    let mut v = Vec::new();
    while n > 0 {
        process_item(n as i32);   // loop edge
        v.push(n);
        n -= 1;
    }
    v
}

/// loop{} body — loop edge.
pub fn retry_until_ok(mut attempts: u32) -> bool {
    loop {
        if validate(attempts as i32) {   // conditional inside loop
            return true;
        }
        if attempts == 0 {
            break;
        }
        attempts -= 1;
    }
    false
}

/// Nested: loop inside if — both loop and conditional present.
pub fn conditional_loop(items: &[i32], threshold: i32) {
    if !items.is_empty() {              // conditional: if-branch
        for &x in items {
            if expensive_check(x) {    // conditional inside loop
                process_item(x);       // loop + conditional
            }
        }
        cleanup();                     // conditional: runs after the if-body
    }
}

/// Ternary-equivalent: if used as expression.
pub fn clamp(x: i32, lo: i32, hi: i32) -> i32 {
    if x < lo { lo }       // conditional
    else if x > hi { hi }  // conditional
    else { x }             // conditional
}

File: fixtures/ts-sample/src/conditions.ts
// cgx-fixture: conditional and loop edges (TypeScript)
// Covers: conditional (if/switch/ternary), loop (for/while/forEach body)

function logInfo(msg: string): void { console.log(`[INFO] ${msg}`); }
function logWarn(msg: string): void { console.warn(`[WARN] ${msg}`); }
function logError(msg: string): void { console.error(`[ERROR] ${msg}`); }
function processItem(x: number): number { return x * 2; }
function cleanup(): void { logInfo("cleanup"); }
function validate(x: number): boolean { return x > 0; }

/// Simple if/else — conditional edges to logInfo, logWarn.
export function maybeLog(flag: boolean): void {
    if (flag) {
        logInfo("flag is true");    // conditional: if-branch
    } else {
        logWarn("flag is false");   // conditional: else-branch
    }
}

/// switch/case — each case produces a conditional edge.
export function dispatch(code: number): string {
    switch (code) {
        case 0:
            logInfo("zero"); return "zero";    // conditional
        case 1:
            logWarn("one"); return "one";      // conditional
        default:
            logError("other"); return "other"; // conditional
    }
}

/// for loop — loop edge inside the body.
export function processAll(items: number[]): number[] {
    const out: number[] = [];
    for (const x of items) {
        out.push(processItem(x));  // loop edge
    }
    return out;
}

/// while loop.
export function countdown(n: number): number[] {
    const v: number[] = [];
    while (n > 0) {
        processItem(n);  // loop edge
        v.push(n);
        n--;
    }
    return v;
}

/// forEach — loop edge inside the callback body.
export function forEachProcess(items: number[]): void {
    items.forEach(x => {
        processItem(x);  // loop edge (inside forEach callback)
    });
}

/// Ternary expression — conditional edges.
export function clamp(x: number, lo: number, hi: number): number {
    return x < lo ? lo : x > hi ? hi : x;  // conditional
}

/// Conditional inside loop.
export function conditionalLoop(items: number[], threshold: number): void {
    if (items.length > 0) {                // conditional
        for (const x of items) {
            if (x > threshold) {           // conditional inside loop
                processItem(x);            // loop + conditional
            }
        }
        cleanup();                         // conditional (after loop, inside outer if)
    }
}

/// Optional chaining — conditional call if value is non-null.
export function maybeValidate(x: number | null): boolean {
    return x !== null && validate(x);  // conditional: only calls validate if x != null
}
